src/app/gibbs/single_thread_sampler.cpp

```diff
@@ -157,25 +157,24 @@ inline int dd::SingleThreadSampler::draw_sample(Variable &variable,
       proposal = -1;
 
       // calculate potential for each proposal
-      for (int propose = variable.lower_bound; propose <= variable.upper_bound;
-           propose++) {
+      for (size_t i = 0; i < variable.cardinality; i++) {
+        int propose = variable.get_domain(i);
         if (is_free_sample) {
-          varlen_potential_buffer[propose] =
+          varlen_potential_buffer[i] =
               p_fg->template potential<true>(variable, propose);
         } else {
-          varlen_potential_buffer[propose] =
+          varlen_potential_buffer[i] =
               p_fg->template potential<false>(variable, propose);
         }
-        sum = logadd(sum, varlen_potential_buffer[propose]);
+        sum = logadd(sum, varlen_potential_buffer[i]);
       }
 
       // flip a coin
       *this->p_rand_obj_buf = erand48(this->p_rand_seed);
-      for (int propose = variable.lower_bound; propose <= variable.upper_bound;
-           propose++) {
-        acc += exp(varlen_potential_buffer[propose] - sum);
+      for (size_t i = 0; i < variable.cardinality; i++) {
+        acc += exp(varlen_potential_buffer[i] - sum);
         if (*this->p_rand_obj_buf <= acc) {
-          proposal = propose;
+          proposal = variable.get_domain(i);
```

Comment:
These loops were the parts that we looked carefully, right?  For dealing with both dense and sparse, how about we keep each body as lambda and just branch into the right loop, like:
```c++
if (variable.domain_map == NULL) {
  for (size_t i = 0; i<variable.cardinality; ++i) body(i, i);
} else {
  for (const auto& entry : variable.domain_map) body(entry.first, entry.second);
}
```